# TASK 3 — Data Cleaning Project: Level 1

## Cleaning and Preparing a Messy Dataset

### Objective

The objective of this project is to systematically clean a messy dataset
and transform it into a clean, consistent, and analysis-ready dataset.

The project includes:
- Data quality assessment
- Missing data handling
- Duplicate removal
- Data standardisation
- Data type correction
- Outlier detection and treatment
- Before vs. after data quality comparison
- Exporting the cleaned dataset

In [ ]:
import pandas as pd
import numpy as np

print("Libraries imported successfully!")

Libraries imported successfully!


In [ ]:
from google.colab import files

uploaded = files.upload()

Saving archive (2).zip to archive (2) (1).zip


### Observation

The Titanic dataset was successfully loaded with 891 rows and 12 columns.
It contains passenger information including survival status, passenger
class, age, gender, ticket, fare, cabin, and embarkation details.

This dataset will be treated as the original "before cleaning" dataset.

In [ ]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

df = pd.read_csv(url)

print("Dataset loaded successfully!")
print("Dataset shape:", df.shape)

Dataset loaded successfully!
Dataset shape: (891, 12)


In [ ]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Observation

The first five records were inspected to understand the structure and
content of the dataset.

The dataset contains both numerical and categorical variables. The initial
inspection also indicates that some columns, such as Age and Cabin, may
contain missing values.

In [ ]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

Number of rows: 891
Number of columns: 12

Column names:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

Data types:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


### Observation

The dataset contains 891 rows and 12 columns.

The numerical columns are stored mainly as integer or floating-point
types, while Name, Sex, Ticket, Cabin, and Embarked are stored as object
types.

The current data types appear reasonable for most columns, but they will
be reviewed and corrected where necessary during the cleaning process.

In [ ]:
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


### Observation

The missing-value analysis identified missing data in three columns:

- Age: 177 missing values
- Cabin: 687 missing values
- Embarked: 2 missing values

Age has a substantial number of missing values, so median imputation will
be considered because Age is a numerical variable and the median is less
affected by extreme values.

Cabin has a very high proportion of missing values. Since replacing most
of the Cabin values would introduce unreliable information, this column
will be considered for removal.

Embarked has only 2 missing values. Since it is a categorical variable,
mode imputation will be used to fill these missing values.

In [ ]:
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


### Observation

The dataset contains 0 duplicate rows.

Therefore, no rows need to be removed during the duplicate-removal step.
The duplicate count will still be documented in the final before-versus-
after quality report.

In [ ]:
df.describe(include='all')

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
count,891.000000,891.000000,891.000000,891,891,714.000000,891.000000,891.000000,891,891.000000,204,889
unique,NaN,NaN,NaN,891,2,NaN,NaN,NaN,681,NaN,147,3
top,NaN,NaN,NaN,"Dooley, Mr. Patrick",male,NaN,NaN,NaN,347082,NaN,G6,S
freq,NaN,NaN,NaN,1,577,NaN,NaN,NaN,7,NaN,4,644
mean,446.000000,0.383838,2.308642,NaN,NaN,29.699118,0.523008,0.381594,NaN,32.204208,NaN,NaN
std,257.353842,0.486592,0.836071,NaN,NaN,14.526497,1.102743,0.806057,NaN,49.693429,NaN,NaN
min,1.000000,0.000000,1.000000,NaN,NaN,0.420000,0.000000,0.000000,NaN,0.000000,NaN,NaN
25%,223.500000,0.000000,2.000000,NaN,NaN,20.125000,0.000000,0.000000,NaN,7.910400,NaN,NaN
50%,446.000000,0.000000,3.000000,NaN,NaN,28.000000,0.000000,0.000000,NaN,14.454200,NaN,NaN
75%,668.500000,1.000000,3.000000,NaN,NaN,38.000000,1.000000,0.000000,NaN,31.000000,NaN,NaN


### Observation

The statistical summary provides an overview of the numerical variables,
including their minimum, maximum, mean, and quartile values.

Variables such as Age, Fare, SibSp, and Parch will be examined further for
potential outliers. The IQR method will be used to identify unusual
numerical observations before deciding whether they should be retained,
capped, or removed.

In [ ]:
quality_report = pd.DataFrame({
    'Column': df.columns,
    'Null Count': df.isnull().sum().values,
    'Data Type': df.dtypes.astype(str).values,
    'Unique Values': df.nunique().values
})

quality_report

,Column,Null Count,Data Type,Unique Values
0,PassengerId,0,int64,891
1,Survived,0,int64,2
2,Pclass,0,int64,3
3,Name,0,object,891
4,Sex,0,object,2
5,Age,177,float64,88
6,SibSp,0,int64,7
7,Parch,0,int64,7
8,Ticket,0,object,681
9,Fare,0,float64,248


### Observation — Initial Data Quality Report

The initial quality report shows that the dataset contains 891 rows and
12 columns.

Missing values were found in three columns:
- Age: 177 missing values
- Cabin: 687 missing values
- Embarked: 2 missing values

There are no missing values in the remaining columns.

The dataset contains both numerical and categorical variables. PassengerId
and Name have 891 unique values, while Ticket and Cabin contain repeated
values.

The current data types appear appropriate for most columns, but they will
be verified during the data type correction stage.

In [ ]:
numeric_columns = df.select_dtypes(include=np.number).columns

range_report = pd.DataFrame({
    'Column': numeric_columns,
    'Minimum': [df[col].min() for col in numeric_columns],
    'Maximum': [df[col].max() for col in numeric_columns],
    'Mean': [df[col].mean() for col in numeric_columns]
})

range_report

,Column,Minimum,Maximum,Mean
0,PassengerId,1.00,891.0000,446.000000
1,Survived,0.00,1.0000,0.383838
2,Pclass,1.00,3.0000,2.308642
3,Age,0.42,80.0000,29.699118
4,SibSp,0.00,8.0000,0.523008
5,Parch,0.00,6.0000,0.381594
6,Fare,0.00,512.3292,32.204208


### Observation — Value Range Analysis

The numerical columns were examined using their minimum, maximum, and
mean values.

The ranges will be compared with the expected meaning of each variable.
For example, Survived should contain only 0 or 1, Pclass should contain
values from 1 to 3, and Age and Fare should contain non-negative values.

Any unusual values identified during this analysis will be investigated
further before the cleaning process.

In [ ]:
before_rows = len(df)
before_duplicates = df.duplicated().sum()
before_nulls = df.isnull().sum().sum()
before_dtypes = df.dtypes.astype(str).to_dict()

print("Before Cleaning")
print("----------------")
print("Rows:", before_rows)
print("Total null values:", before_nulls)
print("Duplicate rows:", before_duplicates)

Before Cleaning
----------------
Rows: 891
Total null values: 866
Duplicate rows: 0


### Observation — Before Cleaning Baseline

The original dataset contains 891 rows and 866 missing values across the
dataset. There are no duplicate rows.

These values are stored as the baseline measurements and will be compared
with the corresponding values after all cleaning operations are completed.

## Missing Data Handling

The missing values were found in three columns: Age, Cabin, and Embarked.

The following strategies are selected:

| Column | Missing Values | Strategy | Reason |
|---|---:|---|---|
| Age | 177 | Median imputation | Age is numerical and the median is less affected by extreme values |
| Cabin | 687 | Drop column | A very large proportion of values are missing, making reliable imputation impractical |
| Embarked | 2 | Mode imputation | Embarked is categorical and only two values are missing |

These decisions are made based on the data type, amount of missing data,
and the meaning of each variable.

In [ ]:
age_median = df['Age'].median()

print("Median Age:", age_median)

Median Age: 28.0


### Observation — Age

The median age is 28.0 years. Since Age is a numerical variable with
missing values, median imputation is selected.

The median is preferred over the mean because it is more robust to
extreme age values and helps preserve the overall distribution of the
dataset.

In [ ]:
df['Age'] = df['Age'].fillna(age_median)

print("Missing Age values after imputation:", df['Age'].isnull().sum())

Missing Age values after imputation: 0


In [ ]:
print("Embarked value counts before imputation:")
print(df['Embarked'].value_counts(dropna=False))

Embarked value counts before imputation:
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


### Observation — Embarked

The Embarked column is a categorical variable with only 2 missing values.

Since only a very small number of records are missing and the column is
categorical, the most frequent category (mode) is an appropriate
replacement strategy.

In [ ]:
embarked_mode = df['Embarked'].mode()[0]

print("Most frequent Embarked value:", embarked_mode)

Most frequent Embarked value: S


In [ ]:
df['Embarked'] = df['Embarked'].fillna(embarked_mode)

print("Missing Embarked values after imputation:",
      df['Embarked'].isnull().sum())

Missing Embarked values after imputation: 0


### Observation — Embarked Imputation

The mode of the Embarked column is "S". The 2 missing values were replaced
with this most frequent category.

This approach is appropriate because Embarked is categorical and only a
very small proportion of its values were missing.

### Observation — Cabin Missing Data

The Cabin column contains 687 missing values out of 891 records,
representing approximately 77% of the column.

Because such a large proportion of the values are missing, mean, median,
mode, or forward-fill imputation would not provide reliable information.

Therefore, the Cabin column will be removed from the dataset. This prevents
the introduction of large amounts of artificial or misleading data.

In [ ]:
df = df.drop(columns=['Cabin'])

print("Cabin column removed successfully.")
print("New dataset shape:", df.shape)

Cabin column removed successfully.
New dataset shape: (891, 11)


In [ ]:
print("Missing values after handling:")
print(df.isnull().sum())

Missing values after handling:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


### Observation — Missing Data After Cleaning

After applying the selected missing-data strategies, the remaining
dataset contains no missing values.

Age missing values were replaced using the median, Embarked missing values
were replaced using the mode, and Cabin was removed because of its very
high proportion of missing data.

The dataset is now ready for the next cleaning stage.

In [ ]:
duplicate_count = df.duplicated().sum()

print("Duplicate rows before removal:", duplicate_count)

Duplicate rows before removal: 0


In [ ]:
df = df.drop_duplicates()

print("Duplicate rows after removal:", df.duplicated().sum())
print("Current dataset shape:", df.shape)

Duplicate rows after removal: 0
Current dataset shape: (891, 11)


### Observation — Duplicate Removal

The dataset was checked for completely duplicated rows, and no duplicate
records were found.

Therefore, no rows were removed during the duplicate-cleaning process.
The dataset remains at 891 rows.

The duplicate count will be included in the final before-versus-after
data quality comparison.

In [ ]:
print("Sex values:")
print(df['Sex'].unique())

print("\nEmbarked values:")
print(df['Embarked'].unique())

Sex values:
['male' 'female']

Embarked values:
['S' 'C' 'Q']


In [ ]:
df['Sex'] = df['Sex'].str.strip().str.capitalize()

print(df['Sex'].unique())

['Male' 'Female']


### Observation — Sex Standardisation

The Sex column originally contained two consistent values: "male" and
"female".

The values were standardised to "Male" and "Female" using whitespace
removal and capitalization. This improves consistency and readability
without changing the underlying meaning of the data.

In [ ]:
df['Embarked'] = df['Embarked'].str.strip().str.upper()

print(df['Embarked'].unique())

['S' 'C' 'Q']


### Observation — Embarked Standardisation

The Embarked column already contained consistent categorical codes:
"S", "C", and "Q".

Whitespace was removed and the values were converted to uppercase to
ensure consistent formatting. No category values were changed.

In [ ]:
text_columns = df.select_dtypes(include='object').columns

for col in text_columns:
    print(f"\n{col}:")
    print(df[col].astype(str).str.strip().head())


Name:
0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

Sex:
0      Male
1    Female
2    Female
3    Female
4      Male
Name: Sex, dtype: object

Ticket:
0           A/5 21171
1            PC 17599
2    STON/O2. 3101282
3              113803
4              373450
Name: Ticket, dtype: object

Embarked:
0    S
1    C
2    S
3    S
4    S
Name: Embarked, dtype: object


### Observation — Text Formatting Check

The text-based columns were checked for unnecessary leading or trailing
whitespace.

Whitespace standardisation helps prevent duplicate categories or
inconsistent text values during future analysis.

In [ ]:
print("Current data types:")
print(df.dtypes)

Current data types:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object


In [ ]:
df['PassengerId'] = df['PassengerId'].astype('string')
df['Name'] = df['Name'].astype('string')
df['Sex'] = df['Sex'].astype('string')
df['Ticket'] = df['Ticket'].astype('string')
df['Embarked'] = df['Embarked'].astype('string')

print(df.dtypes)

PassengerId    string[python]
Survived                int64
Pclass                  int64
Name           string[python]
Sex            string[python]
Age                   float64
SibSp                   int64
Parch                   int64
Ticket         string[python]
Fare                  float64
Embarked       string[python]
dtype: object


### Observation — Data Type Correction

The data types were reviewed based on the meaning of each column.

PassengerId was converted from an integer to a string because it is an
identifier rather than a numerical measurement.

Name, Sex, Ticket, and Embarked were converted to string types because
they contain textual or categorical information.

Age and Fare remain floating-point values because they may contain
decimal values.

Survived, Pclass, SibSp, and Parch remain integer variables because they
represent discrete numerical values.

In [ ]:
print("Data types after correction:")
print(df.dtypes)

Data types after correction:
PassengerId    string[python]
Survived                int64
Pclass                  int64
Name           string[python]
Sex            string[python]
Age                   float64
SibSp                   int64
Parch                   int64
Ticket         string[python]
Fare                  float64
Embarked       string[python]
dtype: object


### Observation — Data Type Verification

The final data types now match the intended meaning of the variables.

Identifiers and textual fields are stored as string values, while
numerical measurements remain numeric. This improves consistency and
reduces the risk of incorrect calculations during analysis.

In [ ]:
numeric_columns = ['Age', 'SibSp', 'Parch', 'Fare']

outlier_report = []

for col in numeric_columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]

    outlier_report.append({
        'Column': col,
        'Q1': Q1,
        'Q3': Q3,
        'IQR': IQR,
        'Lower Bound': lower_bound,
        'Upper Bound': upper_bound,
        'Outlier Count': len(outliers)
    })

outlier_report = pd.DataFrame(outlier_report)

outlier_report

,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,Age,22.0000,35.0,13.0000,2.500,54.5000,66
1,SibSp,0.0000,1.0,1.0000,-1.500,2.5000,46
2,Parch,0.0000,0.0,0.0000,0.000,0.0000,213
3,Fare,7.9104,31.0,23.0896,-26.724,65.6344,116


### Observation — IQR Outlier Detection

The IQR method was used to identify potential outliers in the main
continuous and count-based numerical variables.

An observation is considered a potential outlier when it falls below
Q1 - 1.5 × IQR or above Q3 + 1.5 × IQR.

The identified values will not automatically be removed because an
outlier may represent a valid real-world observation rather than an error.

### Outlier Treatment Decision

The detected outliers will be retained because extreme values in this
dataset can represent genuine passengers or valid ticket fares.

For example, a high Fare value may correspond to a premium ticket, while
higher SibSp or Parch values may represent passengers travelling with
larger families.

Removing these records could result in loss of useful information and
could introduce bias.

Therefore, no valid observations will be removed or capped solely because
they are statistical outliers.

In [ ]:
print("Rows after outlier analysis:", len(df))

Rows after outlier analysis: 891


### Observation — Final Outlier Decision

The outlier analysis was performed using the IQR method. Potential
outliers were identified, but no rows were removed because the extreme
values can represent legitimate passenger characteristics or fares.

The dataset therefore retains all 891 valid records.

In [ ]:
after_rows = len(df)
after_duplicates = df.duplicated().sum()
after_nulls = df.isnull().sum().sum()

print("FINAL DATA QUALITY REPORT")
print("=" * 30)
print("Rows:", after_rows)
print("Total null values:", after_nulls)
print("Duplicate rows:", after_duplicates)
print("Columns:", df.shape[1])

FINAL DATA QUALITY REPORT
Rows: 891
Total null values: 0
Duplicate rows: 0
Columns: 11


In [ ]:
before_after = pd.DataFrame({
    'Metric': [
        'Row Count',
        'Column Count',
        'Total Null Values',
        'Duplicate Rows'
    ],
    'Before Cleaning': [
        before_rows,
        12,
        before_nulls,
        before_duplicates
    ],
    'After Cleaning': [
        after_rows,
        df.shape[1],
        after_nulls,
        after_duplicates
    ]
})

before_after

,Metric,Before Cleaning,After Cleaning
0,Row Count,891,891
1,Column Count,12,11
2,Total Null Values,866,0
3,Duplicate Rows,0,0


### Observation — Before vs After Cleaning

The before-and-after comparison shows the improvements achieved during
the cleaning process.

The dataset originally contained 891 rows, 12 columns, 866 missing
values, and no duplicate rows.

After cleaning, the dataset contains 891 rows and 11 columns with zero
missing values and zero duplicate rows.

The reduction from 12 to 11 columns is due to the removal of Cabin, which
had approximately 77% missing values.

No rows were removed because the existing records were considered valid,
and potential statistical outliers were retained.

In [ ]:
print("FINAL DATA TYPES")
print("=" * 30)

print(df.dtypes)

FINAL DATA TYPES
PassengerId    string[python]
Survived                int64
Pclass                  int64
Name           string[python]
Sex            string[python]
Age                   float64
SibSp                   int64
Parch                   int64
Ticket         string[python]
Fare                  float64
Embarked       string[python]
dtype: object


### Observation — Final Data Types

All columns were reviewed after the cleaning process.

Identifier and textual columns are stored as string types, while
numerical variables such as Age and Fare remain numeric.

The final data types are appropriate for further analysis and machine
learning tasks.

In [ ]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",Male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",Female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",Female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",Female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",Male,35.0,0,0,373450,8.0500,S


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    string 
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    string 
 4   Sex          891 non-null    string 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    string 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    string 
dtypes: float64(2), int64(4), string(5)
memory usage: 76.7 KB


In [ ]:
output_file = "Titanic_Cleaned_Level1.csv"

df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully as:", output_file)

Cleaned dataset saved successfully as: Titanic_Cleaned_Level1.csv


### Observation — Exporting the Cleaned Dataset

The cleaned dataset was successfully exported as
`Titanic_Cleaned_Level1.csv`.

The file contains the cleaned, standardised, and analysis-ready data and
can be used for further data analysis or machine learning tasks.

In [ ]:
from google.colab import files

files.download("Titanic_Cleaned_Level1.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Conclusion

The Titanic dataset was systematically cleaned using Python, Pandas, and
NumPy.

The cleaning process included:

- Initial data quality assessment
- Missing-value analysis and treatment
- Median imputation for Age
- Mode imputation for Embarked
- Removal of Cabin due to excessive missing values
- Duplicate detection and removal
- Categorical data standardisation
- Data type correction
- IQR-based outlier detection
- Retention of valid statistical outliers
- Before-versus-after quality comparison
- Exporting the final cleaned dataset

The final dataset contains 891 rows and 11 columns with no missing values
or duplicate rows. The resulting dataset is clean, consistent, and ready
for further analysis.